text/3723-static-function-argument-unpacking.md

```diff
@@ -0,0 +1,753 @@
+
+Rust GitHub:
+- [Draft RFC: variadic generics](https://github.com/rust-lang/rfcs/issues/376)
+- [RFC 2909: destructuring_assignment](https://github.com/rust-lang/rfcs/blob/master/text/2909-destructuring-assignment.md)
+
+Rust Internals:
+- [Variadic generics design sketch](https://internals.rust-lang.org/t/variadic-generics-design-sketch/18974)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should argument unpacking desugar into Alternative A or Alternative B below, or does it make any difference?
+    
+    Alternative A:
+    ```rust
+    let tup = (1, 2, 3);
+    
+    // callee_fn(...tup); desugars into:
+    {
+        let _tmp0 = tup.0;
+        let _tmp1 = tup.1;
+        let _tmp2 = tup.2;
+        callee_fn(_tmp0, _tmp1, _tmp2);
+    }
+    ```
+    
+    Alternative B:
+    ```rust
+    let tup = (1, 2, 3);
+    
+    // callee_fn(...tup); desugars into:
+    callee_fn(tup.0, tup.1, tup.2);
+    ```
+
+- What happens when unpacking an empty collection as the "arguments" for a function that doesn't define parameters?
+    - Probably nothing, at least if functions without parameters are not special-cased. Experimenting with this will be easier with a proof-of-concept implementation.
+
+# Future Possibilities
+[future-possibilities]: #future-possibilities
+
+Future RFCs should freely venture outside the scope of this RFC and complement or build on this limited form of argument unpacking.
+
+## Unpacking in Fixed-Size Array and Tuple Literals
+
+The same ellipsis syntax with a very similar meaning could be adopted to defining fixed-size arrays and tuple literals as well. For example:
+```rust
+const CDE: [char; 3] = ['C', 'D', 'E'];
+const ABCDEFG1: [char; 7] = ['A', 'B', ...CDE, 'F', 'G'];
+const ABCDEFG2: [char; 7] = ['A', 'B', CDE[0], CDE[1], CDE[2], 'F', 'G'];
+
+assert_eq(ABCDEFG1, ABCDEFG2);
+```
+
+At least Python and JavaScript have this feature.
+
+Notably, tuple structs can already be built with the adoption of argument unpacking in general, since their constructors use the call expression that is modified by the addition of argument unpacking.
+
+Prior work on designing such feature exists at least in the Rust Internals thread [Pre-RFC: Array expansion syntax](https://internals.rust-lang.org/t/pre-rfc-array-expansion-syntax/13490).
+
+## Unpacking Arguments for Macro Invocations
+
+Macros, callable with the `macro_name!(…)`, `macro_name![…]`, or `macro_name!{…}` syntax have been omitted from the scope of this proposal. The reason for omission is the time concerns related to doing the due diligence researching the design. For example, some macros (e.g. `println!`) accept an indefinite number of arguments.
```

Comment:
why is handling macro invocation is even considered in this RFC lol

the input to a macro always see `...expr` as two tokens and it should never change

```rust
macro_rules! backward_compatiblity_check {
    ($($t:tt)*) => { println!(concat!($("<", stringify!($t), "> "),*)) }
}

fn main() {
    backward_compatiblity_check![1, ...expr, 2];
    // <1> <,> <...> <expr> <,> <2> 
}
```

if `vec![]` or `println!()` want to support unpacking it is the responsibility of the implementation of those macros themselves individually.

Reply:
I've clarified the "Future Possibilities" subchapter "Unpacking Arguments for Macro Invocations" in commit e2ec5b10132256b976d08b2fc6ec314d4a03fcd9. This change would presumably be on the turf of the library team, and I think discussing its postponement to a separate proposal is in line with this RFC – specifically, in getting a minimum useful version of the feature into the language first, and then expanding on that with further initiatives.

Anyway, thanks for the declarative macro example – it helped me clear up some token-related stuff!